File: decoder-cpp/src/EditDistance.cpp
//
// Created by augustus on 8/30/18.
//

#include "../include/EditDistance.h"
#include <vector>


EditDistance::EditDistance() {

}

EditDistance::~EditDistance() {

}

int EditDistance::edit_distance(string &left, string &right) {
    if(left.empty()) return (int)right.size();
    if(right.empty()) return (int)left.size();
    int data[left.size()+1][right.size()+1];
    for(int i=0; i<right.size()+1; i++){
        data[0][i] = i;
    }
    for(int j=0; j<left.size()+1; j++){
        data[j][0] = j;
    }
    for(int m=0; m<left.size(); m++){
        for(int n=0; n < right.size(); n++){
            if(left[m]==right[n]){
                data[m+1][n+1] = data[m][n];
            }
            else{
                int a = data[m+1][n] + 1;
                int b = data[m][n+1] + 1;
                int c = data[m][n] + 1;
                data[m+1][n+1] = min(a, min(b, c));
            }
        }
    }
    return data[left.size()][right.size()];
}

File: decoder-python/src/Editdistance.py
"""
Library used to calculate edit-distance between two strings
"""
import numpy as np


def edit_distance(left: str, right: str)->int:
    """
    Quick function to calculate edit-distacne
    :param left: string one
    :param right: string two
    :return: edit distance
    """
    if left is None:
        return len(right)
    if right is None:
        return len(left)

    dis = np.zeros((len(left)+1, len(right)+1))
    for m in range(len(right)+1):
        dis[0][m] = m
    for n in range(len(left)+1):
        dis[n][0] = n
    for i in range(len(left)):
        for j in range(len(right)):
            if left[i] == right[j]:
                dis[i+1][j+1] = dis[i][j]
            else:
                a = dis[i][j+1] + 1
                b = dis[i+1][j] + 1
                c = dis[i][j] + 1
                dis[i+1][j+1] = min(a, min(b, c))

    return dis[len(left)][len(right)]


if __name__ == "__main__":
    test_l = "FITNESSSYSTEM"
    test_r = 'FITNESSSYSTEM'
    res = edit_distance(test_l, test_r)
    print("Stop")
